airbyte-cdk/bulk/toolkits/load-avro/src/main/kotlin/io/airbyte/cdk/load/data/avro/AirbyteValueToAvroRecord.kt

```diff
@@ -4,73 +4,184 @@
-            is TimeValue ->
-                throw IllegalArgumentException("String-based time types are not supported")
-            is TimestampValue ->
-                throw IllegalArgumentException("String-based timestamp types are not supported")
-            is UnknownValue -> throw IllegalArgumentException("Unknown type is not supported")
+            BooleanType -> return (airbyteValue as BooleanValue).value
+            DateType -> return (airbyteValue as IntegerValue).value.toInt()
+            IntegerType -> return (airbyteValue as IntegerValue).value.toLong()
+            NumberType -> return (airbyteValue as NumberValue).value.toDouble()
+            StringType -> return (airbyteValue as StringValue).value
+
+            // Converted to strings upstream
+            is UnknownType,
+            ObjectTypeWithEmptySchema,
+            ObjectTypeWithoutSchema,
+            ArrayTypeWithoutSchema -> return (airbyteValue as StringValue).value
+
+            // Converted to integrals upstream
+            TimeTypeWithTimezone,
+            TimeTypeWithoutTimezone,
+            TimestampTypeWithTimezone,
+            TimestampTypeWithoutTimezone -> return (airbyteValue as IntegerValue).value.toLong()
+            is UnionType -> {
+                for (optionType in airbyteSchema.options) {
+                    try {
+                        val optionSchema = matchingAvroType(optionType, schema)
+                        return convert(airbyteValue, optionType, optionSchema)
+                    } catch (e: Exception) {
+                        continue
+                    }
+                }
+                throw IllegalArgumentException(
+                    "No matching Avro type found for $airbyteSchema in $schema (airbyte value: ${airbyteValue.javaClass.simpleName})"
+                )
+            }
+        }
+    }
+
+    private fun matchingAvroType(airbyteSchema: AirbyteType, avroUnionSchema: Schema): Schema {
+        return when (airbyteSchema) {
+            is ObjectType -> avroUnionSchema.types.find { it.type == Schema.Type.RECORD }
+            is ArrayType -> avroUnionSchema.types.find { it.type == Schema.Type.ARRAY }
+            BooleanType -> avroUnionSchema.types.find { it.type == Schema.Type.BOOLEAN }
+            DateType -> avroUnionSchema.types.find { it.type == Schema.Type.INT }
+            IntegerType -> avroUnionSchema.types.find { it.type == Schema.Type.LONG }
+            NumberType -> avroUnionSchema.types.find { it.type == Schema.Type.DOUBLE }
+            is UnknownType,
+            ArrayTypeWithoutSchema,
+            ObjectTypeWithEmptySchema,
+            ObjectTypeWithoutSchema,
+            StringType -> avroUnionSchema.types.find { it.type == Schema.Type.STRING }
+            TimeTypeWithTimezone,
+            TimeTypeWithoutTimezone,
+            TimestampTypeWithTimezone,
+            TimestampTypeWithoutTimezone ->
+                avroUnionSchema.types.find { it.type == Schema.Type.LONG }
```

Comment:
Is this right? Do we convert the field to a timestamp?

Reply:
Yep, avro wants integral time types